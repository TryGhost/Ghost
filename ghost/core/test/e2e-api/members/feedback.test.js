const assert = require('assert/strict');
const {agentProvider, mockManager, fixtureManager, matchers, configUtils} = require('../../utils/e2e-framework');
const {anyEtag, anyObjectId, anyLocationFor, anyErrorId} = matchers;
const models = require('../../../core/server/models');
const sinon = require('sinon');
const settingsHelpers = require('../../../core/server/services/settings-helpers');
const crypto = require('crypto');

const membersValidationKeyMock = 'abc123dontstealme';

describe('Members Feedback', function () {
    let membersAgent, membersAgent2, memberUuid, memberHmac;
    let clock;

    before(async function () {
        membersAgent = await agentProvider.getMembersAPIAgent();
        membersAgent2 = membersAgent.duplicate();

        await fixtureManager.init('posts', 'members');
        memberUuid = fixtureManager.get('members', 0).uuid;
        memberHmac = crypto.createHmac('sha256', membersValidationKeyMock).update(memberUuid).digest('hex');
    });

    beforeEach(function () {
        sinon.stub(settingsHelpers, 'getMembersValidationKey').returns(membersValidationKeyMock);
        mockManager.mockMail();
    });

    afterEach(async function () {
        clock?.restore();
        clock = undefined;
        await configUtils.restore();
        mockManager.restore();
    });

    describe('Authentication', function () {
        it('Allows authentication via session', async function () {
            const postId = fixtureManager.get('posts', 0).id;
            await membersAgent2.loginAs('authenticationtest@email.com');

            await membersAgent2
                .post('/api/feedback/')
                .body({
                    feedback: [{
                        score: 1,
                        post_id: postId
                    }]
                })
                .expectStatus(201)
                .matchHeaderSnapshot({
                    etag: anyEtag,
                    location: anyLocationFor('feedback')
                })
                .matchBodySnapshot({
                    feedback: [
                        {
                            id: anyObjectId,
                            memberId: anyObjectId,
                            postId: anyObjectId
                        }
                    ]
                });
        });

        it('Allows authentication via uuid (+ key)', async function () {
            const postId = fixtureManager.get('posts', 0).id;

            await membersAgent
                .post(`/api/feedback/?uuid=${memberUuid}&key=${memberHmac}`)
                .body({
                    feedback: [{
                        score: 1,
                        post_id: postId
                    }]
                })
                .expectStatus(201)
                .matchHeaderSnapshot({
                    etag: anyEtag,
                    location: anyLocationFor('feedback')
                })
                .matchBodySnapshot({
                    feedback: [
                        {
                            id: anyObjectId,
                            memberId: anyObjectId,
                            postId: anyObjectId
                        }
                    ]
                });
        });

        it('Throws for missing key', async function () {
            const postId = fixtureManager.get('posts', 0).id;
            await membersAgent
                .post(`/api/feedback/?uuid=${memberUuid}`)
                .body({
                    feedback: [{
                        score: 1,
                        post_id: postId
                    }]
                })
                .expectStatus(401)
                .matchBodySnapshot({
                    errors: [
                        {
                            id: anyErrorId
                        }
                    ]
                });
        });

        it('Thorws for invalid key', async function () {
            const postId = fixtureManager.get('posts', 0).id;
            await membersAgent
                .post(`/api/feedback/?uuid=${memberUuid}&key=1234`)
                .body({
                    feedback: [{
                        score: 1,
                        post_id: postId
                    }]
                })
                .expectStatus(401)
                .matchBodySnapshot({
                    errors: [
                        {
                            id: anyErrorId
                        }
                    ]
                });
        });

        it('Throws for invalid uuid', async function () {
            const postId = fixtureManager.get('posts', 0).id;
            await membersAgent
                .post(`/api/feedback/?uuid=1234`)
                .body({
                    feedback: [{
                        score: 1,
                        post_id: postId
                    }]
                })
                .expectStatus(401)
                .matchBodySnapshot({
                    errors: [
                        {
                            id: anyErrorId
                        }
                    ]
                });
        });

        it('Throws for nonexisting uuid', async function () {
            const postId = fixtureManager.get('posts', 0).id;
            const uuid = '00000000-0000-0000-0000-000000000000';
            await membersAgent
                .post(`/api/feedback/?uuid=${uuid}`)
                .body({
                    feedback: [{
                        score: 1,
                        post_id: postId
                    }]
                })
                .expectStatus(401)
                .matchBodySnapshot({
                    errors: [
                        {
                            id: anyErrorId
                        }
                    ]
                });
        });
    });

    describe('Validation', function () {
        const postId = fixtureManager.get('posts', 0).id;

        it('Throws for invalid score', async function () {
            await membersAgent
                .post(`/api/feedback/?uuid=${memberUuid}&key=${memberHmac}`)
                .body({
                    feedback: [{
                        score: 2,
                        post_id: postId
                    }]
                })
                .expectStatus(422)
                .matchBodySnapshot({
                    errors: [
                        {
                            id: anyErrorId
                        }
                    ]
                });
        });

        it('Throws for invalid score type', async function () {
            await membersAgent
                .post(`/api/feedback/?uuid=${memberUuid}&key=${memberHmac}`)
                .body({
                    feedback: [{
                        score: 'text',
                        post_id: postId
                    }]
                })
                .expectStatus(422)
                .matchBodySnapshot({
                    errors: [
                        {
                            id: anyErrorId
                        }
                    ]
                });
        });

        it('Throws for nonexisting post', async function () {
            await membersAgent
                .post(`/api/feedback/?uuid=${memberUuid}&key=${memberHmac}`)
                .body({
                    feedback: [{
                        score: 1,
                        post_id: '123'
                    }]
                })
                .expectStatus(404)
                .matchBodySnapshot({
                    errors: [
                        {
                            id: anyErrorId
                        }
                    ]
                });
        });
    });

    it('Can add positive feedback', async function () {
        const postId = fixtureManager.get('posts', 0).id;

        await membersAgent
            .post(`/api/feedback/?uuid=${memberUuid}&key=${memberHmac}`)
            .body({
                feedback: [{
                    score: 1,
                    post_id: postId
                }]
            })
            .expectStatus(201)
            .matchHeaderSnapshot({
                etag: anyEtag,
                location: anyLocationFor('feedback')
            })
            .matchBodySnapshot({
                feedback: [
                    {
                        id: anyObjectId,
                        memberId: anyObjectId,
                        postId: anyObjectId
                    }
                ]
            });
    });

    it('Can change existing feedback', async function () {
        clock = sinon.useFakeTimers(new Date());
        const postId = fixtureManager.get('posts', 1).id;

        const {body} = await membersAgent
            .post(`/api/feedback/?uuid=${memberUuid}&key=${memberHmac}`)
            .body({
                feedback: [{
                    score: 0,
                    post_id: postId
                }]
            })
            .expectStatus(201)
            .matchHeaderSnapshot({
                etag: anyEtag,
                location: anyLocationFor('feedback')
            })
            .matchBodySnapshot({
                feedback: [
                    {
                        id: anyObjectId,
                        memberId: anyObjectId,
                        postId: anyObjectId
                    }
                ]
            });

        assert.equal(body.feedback[0].score, 0);
        const feedbackId = body.feedback[0].id;

        // Fetch real model
        const model1 = await models.MemberFeedback.findOne({id: feedbackId}, {require: true});

        clock.tick(10 * 60 * 1000);

        const {body: body2} = await membersAgent
            .post(`/api/feedback/?uuid=${memberUuid}&key=${memberHmac}`)
            .body({
                feedback: [{
                    score: 1,
                    post_id: postId
                }]
            })
            .expectStatus(201)
            .matchHeaderSnapshot({
                etag: anyEtag,
                location: anyLocationFor('feedback')
            })
            .matchBodySnapshot({
                feedback: [
                    {
                        id: anyObjectId,
                        memberId: anyObjectId,
                        postId: anyObjectId
                    }
                ]
            });

        assert.equal(body2.feedback[0].id, feedbackId);
        assert.equal(body2.feedback[0].score, 1);
        const model2 = await models.MemberFeedback.findOne({id: feedbackId}, {require: true});

        assert.notEqual(model2.get('updated_at').getTime(), model1.get('updated_at').getTime());

        clock.tick(10 * 60 * 1000);

        // Do the same change again, and the model shouldn't change
        const {body: body3} = await membersAgent
            .post(`/api/feedback/?uuid=${memberUuid}&key=${memberHmac}`)
            .body({
                feedback: [{
                    score: 1,
                    post_id: postId
                }]
            })
            .expectStatus(201)
            .matchHeaderSnapshot({
                etag: anyEtag,
                location: anyLocationFor('feedback')
            })
            .matchBodySnapshot({
                feedback: [
                    {
                        id: anyObjectId,
                        memberId: anyObjectId,
                        postId: anyObjectId
                    }
                ]
            });

        const model3 = await models.MemberFeedback.findOne({id: feedbackId}, {require: true});
        assert.equal(body3.feedback[0].id, feedbackId);
        assert.equal(body3.feedback[0].score, 1);
        assert.equal(model3.get('updated_at').getTime(), model2.get('updated_at').getTime());
    });
});
