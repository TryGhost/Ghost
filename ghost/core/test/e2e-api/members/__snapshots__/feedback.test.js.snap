// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Members Feedback Authentication Allows authentication via session 1: [body] 1`] = `
Object {
  "feedback": Array [
    Object {
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "memberId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "postId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "score": 1,
    },
  ],
}
`;

exports[`Members Feedback Authentication Allows authentication via session 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "132",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": StringMatching /https\\?:\\\\/\\\\/\\.\\*\\?\\\\/feedback\\\\/\\[a-f0-9\\]\\{24\\}\\\\//,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Members Feedback Authentication Allows authentication via uuid (+ key) 1: [body] 1`] = `
Object {
  "feedback": Array [
    Object {
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "memberId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "postId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "score": 1,
    },
  ],
}
`;

exports[`Members Feedback Authentication Allows authentication via uuid (+ key) 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "132",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": StringMatching /https\\?:\\\\/\\\\/\\.\\*\\?\\\\/feedback\\\\/\\[a-f0-9\\]\\{24\\}\\\\//,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Members Feedback Authentication Thorws for invalid key 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": null,
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Invalid key.",
      "property": null,
      "type": "UnauthorizedError",
    },
  ],
}
`;

exports[`Members Feedback Authentication Throws for invalid uuid 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": null,
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Invalid key.",
      "property": null,
      "type": "UnauthorizedError",
    },
  ],
}
`;

exports[`Members Feedback Authentication Throws for missing key 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": null,
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Invalid key.",
      "property": null,
      "type": "UnauthorizedError",
    },
  ],
}
`;

exports[`Members Feedback Authentication Throws for nonexisting uuid 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": null,
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Invalid key.",
      "property": null,
      "type": "UnauthorizedError",
    },
  ],
}
`;

exports[`Members Feedback Can add feedback 1: [body] 1`] = `
Object {
  "feedback": Array [
    Object {
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "memberId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "postId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "type": "positive",
    },
  ],
}
`;

exports[`Members Feedback Can add feedback 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "140",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": StringMatching /https\\?:\\\\/\\\\/\\.\\*\\?\\\\/feedback\\\\/\\[a-f0-9\\]\\{24\\}\\\\//,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Members Feedback Can add positive feedback 1: [body] 1`] = `
Object {
  "feedback": Array [
    Object {
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "memberId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "postId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "score": 1,
    },
  ],
}
`;

exports[`Members Feedback Can add positive feedback 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "132",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": StringMatching /https\\?:\\\\/\\\\/\\.\\*\\?\\\\/feedback\\\\/\\[a-f0-9\\]\\{24\\}\\\\//,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Members Feedback Can change existing feedback 1: [body] 1`] = `
Object {
  "feedback": Array [
    Object {
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "memberId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "postId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "score": 0,
    },
  ],
}
`;

exports[`Members Feedback Can change existing feedback 2: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "132",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": StringMatching /https\\?:\\\\/\\\\/\\.\\*\\?\\\\/feedback\\\\/\\[a-f0-9\\]\\{24\\}\\\\//,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Members Feedback Can change existing feedback 3: [body] 1`] = `
Object {
  "feedback": Array [
    Object {
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "memberId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "postId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "score": 1,
    },
  ],
}
`;

exports[`Members Feedback Can change existing feedback 4: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "132",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": StringMatching /https\\?:\\\\/\\\\/\\.\\*\\?\\\\/feedback\\\\/\\[a-f0-9\\]\\{24\\}\\\\//,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Members Feedback Can change existing feedback 5: [body] 1`] = `
Object {
  "feedback": Array [
    Object {
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "memberId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "postId": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "score": 1,
    },
  ],
}
`;

exports[`Members Feedback Can change existing feedback 6: [headers] 1`] = `
Object {
  "access-control-allow-origin": "*",
  "cache-control": "no-cache, private, no-store, must-revalidate, max-stale=0, post-check=0, pre-check=0",
  "content-length": "132",
  "content-type": "application/json; charset=utf-8",
  "etag": StringMatching /\\(\\?:W\\\\/\\)\\?"\\(\\?:\\[ !#-\\\\x7E\\\\x80-\\\\xFF\\]\\*\\|\\\\r\\\\n\\[\\\\t \\]\\|\\\\\\\\\\.\\)\\*"/,
  "location": StringMatching /https\\?:\\\\/\\\\/\\.\\*\\?\\\\/feedback\\\\/\\[a-f0-9\\]\\{24\\}\\\\//,
  "vary": "Accept-Encoding",
  "x-powered-by": "Express",
}
`;

exports[`Members Feedback Validation Throws for invalid score 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Invalid feedback score. Only 1 or 0 is currently allowed.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save feedback.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Members Feedback Validation Throws for invalid score type 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Invalid feedback score. Only 1 or 0 is currently allowed.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save feedback.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Members Feedback Validation Throws for invalid type 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Invalid feedback type",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save feedback.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Members Feedback Validation Throws for invalid uuid 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": null,
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Invalid uuid.",
      "property": null,
      "type": "UnauthorizedError",
    },
  ],
}
`;

exports[`Members Feedback Validation Throws for nonexisting post 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Post not found.",
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Resource not found error, cannot save feedback.",
      "property": null,
      "type": "NotFoundError",
    },
  ],
}
`;

exports[`Members Feedback Validation Throws for nonexisting uuid 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": null,
      "details": null,
      "ghostErrorCode": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Invalid uuid.",
      "property": null,
      "type": "UnauthorizedError",
    },
  ],
}
`;
